origin-growth/src/resources/authentication.js

```diff
@@ -0,0 +1,126 @@
+  'I accept the terms of growth campaign version: 1.0'
+
+/**
+ * Authenticates user's enrollment to the growth campaign
+ * @param {string} referrer - Eth address of the referrer.
+ * @param {Array<string>>} recipients - List of email addresses.
+ */
+async function authenticateEnrollment(accountId, agreementMessage, signature) {
+  if (currentAgreementMessage !== agreementMessage) {
+    throw new Error(
+      `Incorrect agreementMessage. Expected: "${currentAgreementMessage}" received: "${agreementMessage}"`
+    )
+  }
+  const recoveredAccountId = web3.eth.accounts.recover(
+    agreementMessage,
+    signature
+  )
+
+  if (accountId !== recoveredAccountId) {
+    throw new Error('Recovered and provided accounts do not match')
+  }
+
+  const participant = await db.GrowthParticipant.findOne({
+    where: {
+      ethAddress: accountId
+    }
+  })
+
+  if (
+    participant !== null &&
+    participant.status === enums.GrowthParticipantStatuses.Banned
+  ) {
+    logger.warn(`Banned user: ${accountId} tried to enroll`)
+    throw new Error('This user is banned')
+  }
+
+  const authToken =
+    participant === null
+      ? crypto.randomBytes(64).toString('hex')
+      : participant.authToken
+
+  const participantData = {
+    ethAddress: accountId,
+    status: enums.GrowthParticipantStatuses.Active,
+    agreementId: agreementMessage,
+    // if user uses growth from 2 devices let them share the same auth token
+    authToken: authToken
+  }
+
+  if (participant !== null) {
+    await participant.update(participantData)
+    logger.info(`Existing user enrolled into growth campaign: ${accountId}`)
+  } else {
+    await db.GrowthParticipant.create(participantData)
+    logger.info(`New user enrolled into growth campaign: ${accountId}`)
+  }
+
+  await createInviteCode(accountId)
+  return authToken
+}
+
+/**
+ * Fetches user's authentication status
+ * @param {string} token - Growth authentication token
+ *
+ * returns GrowthParticipantAuthenticationStatus
+ *  - Enrolled -> user participates in growth campaign
+ *  - Banned -> user is banned
+ *  - NotEnrolled -> user not a participant yet
+ */
+async function getUserAuthenticationStatus(token) {
+  const growthParticipant = await db.GrowthParticipant.findOne({
+    where: {
+      authToken: token
+    }
+  })
+
+  if (growthParticipant === null) {
+    return enums.GrowthParticipantAuthenticationStatus.NotEnrolled
+  } else if (
+    growthParticipant.status === enums.GrowthParticipantStatuses.Banned
+  ) {
+    return enums.GrowthParticipantAuthenticationStatus.Banned
+  } else {
+    return enums.GrowthParticipantAuthenticationStatus.Enrolled
+  }
+}
+
+async function createInviteCode(accountId) {
+  const existingInvite = await db.GrowthInviteCode.findOne({
+    where: {
+      ethAddress: accountId
+    }
+  })
+
+  // don't override the existing invite code
+  if (existingInvite !== null) {
+    return
+  }
+
+  const code =
+    `${accountId.substring(2, 5)}-` +
```

Comment:
How are ETH addresses generated. Don't we have a risk of collision ? Granted it is pretty small, but still a risk ?

One idea to generate unique invite code would be to leverage GrowthInviteCode.id since it is a sequence and will generate unique ids. It would go like this:
 - insert an entry in GrowthInviteCode for eth_address.
 - get the id
 - generate a unique code based on the id by using symmetric encryption on the id or some sort of hashing. we could use the id directly as code, but perhaps we don't want to expose it in the clear since it would reveal how many participants in the program ?
 - update GrowthInviteCode with the generated code.

Another possibility is to keep what you have but check for collision by trying to load any existing row that may have the same code. If found, the modify slightly the code and try again.

Reply:
Very nice catch thanks! When creating the growth invite code I now get the count of `GrowthInviteCode` table and convert that to human friendly hash, that represent the last characters of the invite code.